### PR TITLE
Diff.py: Add delete and rename functionalities

### PR DESCRIPTION
**Pull request by arush0311**
This adds rename and delete properties to Diff.py class

Fixes https://github.com/coala-analyzer/coala/issues/1419

### CODE REVIEW

**Comment by srisankethu** [comment]
@arush0311 Don't use '!' for negation. Use 'not' instead. Python doesn't use '!'. Should have run your patch before submitting.

**Comment by SanketDG** [comment]
@arush0311 You will want to cover all added lines in the tests. See this for reference:
https://codecov.io/github/coala-analyzer/coala/coalib/results/Diff.py?ref=7ab8c68cae954019a2f0396a844883fba4fa41fe

**Comment by arush0311** [comment]
ok

**Comment by arush0311** [comment]
@abhsag24 Can you tell me why code coverage is failing?

**Comment by abhsag24** [comment]
@arush0311 also, you could always ask for assistance on our [gitter channel](https://gitter.im/coala-analyzer/coala)  :)

**Comment by arush0311** [comment]
Thanks

**Comment by AbdealiJK** [comment]
This looks good to me at first glance (other than minor errors)

But Im wondering what happens when `__add__` is called with a "deleted" or "renamed" Diff.
Also, I remember we have `overlaps` in some parts of our code. Would that get affected by this ?

Would be awesome if someone else can co-review it.

**Comment by sils1297** [comment]
unack 44d41dd but it's solid start :) I guess this may need a few iterations more, this affects some core behaviour and we need to be careful with that.

**Comment by Adrianzatreanu** [comment]
hey @arush0311 , you still working on this? long time since your last activity on this PR :)

**Comment by arush0311** [comment]
@Adrianzatreanu yes. Please give me some more time to work on this.

**Comment by Adrianzatreanu** [comment]
Dont worry about time. I was just messaging you to see if you're still here, glad to hear you are! Some people leave sometimes, thats why i felt like pinging you :) we're still on `coala <https://gitter.im/coala-analyzer/coala>`_ if you need help!

**Comment by sils1297** [comment]
Hey if you need any help ping us :)

File: coalib/results/Diff.py [hunk 1]
```diff
@@ -257,4 +274,5 @@ def change_line(self, line_nr, original_line, replacement):
 
     def __eq__(self, other):
         return ((self._file == other._file) and
-                (self.modified == other.modified))
+                (self.modified == other.modified) and
+                (self.rename == other.rename))
```
Comment (sils1297) [inline coalib/results/Diff.py]: hey we have a generate_eq function in Decorators.py could you do another commit on top using that one?

File: coalib/results/Diff.py [hunk 2]
```diff
@@ -23,6 +24,7 @@ def __init__(self, file_list):
         """
         self._changes = {}
         self._file = file_list
+        self._rename = False
```
Comment (sils1297) [inline coalib/results/Diff.py]: why not add an optional argument to init?

File: tests/results/DiffTest.py [hunk 3]
```diff
@@ -191,3 +195,11 @@ def test_rename(self):
...
         with self.assertRaises(TypeError):
             self.uut.rename = 1234
+
+    def test_delete(self):
+        self.uut.delete = True
+        self.uut.delete = False
+        # Double deletion is allowed
+        self.uut.delete = False
+        with self.assertRaises(TypeError):
+            self.uut.delete = "abcd"
```
Comment (sils1297) [inline tests/results/DiffTest.py]: you don't test modified here d you?

**Comment by sils1297** [comment]
hm so one important part is that the ApplyPatchAction and the ShowPatchAction need to be modified to act accordingly, in an own commit each

File: tests/results/DiffTest.py (hunk 3 above)
Comment (arush0311) [inline tests/results/DiffTest.py]: No. I wonder how code coverage is showing a 100%

**Comment by sils1297** [comment]
hey this ons needs a rebase, it even has conflcts apparently

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/1801/commits/4b092d3c2c5f2014f05e193ffd05dd56f4606e6f https://github.com/coala-analyzer/coala/pull/1801/commits/5bd3cc4fb43ccb6b43a7aefaa93e6ea306c01b15 https://github.com/coala-analyzer/coala/pull/1801/commits/e5b789a26d78962477d1b939bb5dee6c1c0e12df

**Comment by sils1297** [comment]
alright @arush0311 this is good, it needs two things to be merged:

- a rebase
- the `unified_diff` method should receive some love to show `/dev/null` or something for a removed file and mirror renamings?
- the ApplyPatchAction and ShowPatchAction both need a commit to handle renamings and deletions properly

**Comment by arush0311** [comment]
Actually these are 3 things :p

**Comment by sils1297** [comment]
oh yeah a third one came up while I was writing the comment :P

**Comment by Adrianzatreanu** [comment]
this seems to need a rebase :)

**Comment by Adrianzatreanu** [comment]
hey @arush0311 could you rebase and do this? It seems to have been acked and passed the tests, we might merge this :)

**Comment by sils1297** [comment]
@Adrianzatreanu this isn't ready, the commits are but some stuff needs to be done still

**Comment by Adrianzatreanu** [comment]
i understand then

**Comment by SanketDG** [comment]
@arush0311 Ping. Are you still working on this? All you need is a rebase! :)

**Comment by srisankethu** [comment]
Hey! @arush0311 Looks like your patch is near completion. Few corrections and rebase. Please ping us if you have an issues.

**Comment by sils1297** [comment]
@arush0311 I'm taking this over, want this for the release

**Comment by sils1297** [comment]
improved version on https://github.com/coala-analyzer/coala/pull/2077
